### DISCUSSION TITLE
Raspberry Pi Zero W issues

### DISCUSSION THREAD

**Discussion by toreupfeet** [mKeRix/room-assistant#453] [Q&A]
I am trying to get Room Assistant to work on a Pi Zero W. Whenever I include BLE in my config, there is an `Error: ENODEV, No such device`. I'll post my config and then my logs. Even when I don't include BLE, I have a separate issue. But for now I'll just post this BLE issue in hopes someone will take a look.

```
  hciDeviceId:
    - 0
```

...worked for another user in these threads, but I'm still having issues.

```
global:
  instanceName: pi_zero_front_door
  integrations:
    - homeAssistant
    - bluetoothClassic
    - bluetoothLowEnergy
homeAssistant:
  mqttUrl: 'mqtt://192.168.50.100:1883'
  mqttOptions:
    username: USER
    password: PASS
bluetoothClassic:
  updatefrequency: 1
  maxdistance: 5
  minRssi: -2
  hciDeviceId:
    - 1
  addresses:
    - 'f0:99:bf:28:3a:50'
    - 'e0:2b:96:d5:67:08'
    - 'd8:dc:40:89:0d:6a'
bluetoothLowEnergy:
  hciDeviceId:
    - 0
  updateFrequency: 1
  maxDistance: 5
  whitelist:
    - d5202fc6ff24
  ```

Log

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[01:56:11] INFO: Setting up Home Assistant configuration
[01:56:11] INFO: Starting room-assistant
*** WARNING *** The program 'node' uses the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/blog/projects/avahi-compat.html>
*** WARNING *** The program 'node' called 'DNSServiceRegister()' which is not supported (or only supported partially) in the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/blog/projects/avahi-compat.html>
1/16/2021, 1:56:13 AM - info - IntegrationsModule: Loading integrations: home-assistant, bluetooth-classic, bluetooth-low-energy
1/16/2021, 1:56:14 AM - info - NestFactory: Starting Nest application...
1/16/2021, 1:56:14 AM - info - InstanceLoader: AppModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: ConfigModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: NestEmitterModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: IntegrationsModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: HttpModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: DiscoveryModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: ClusterModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: TerminusModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: BluetoothModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: ScheduleModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: EntitiesModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: HomeAssistantModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: BluetoothLowEnergyModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: BluetoothClassicModule dependencies initialized
1/16/2021, 1:56:14 AM - info - InstanceLoader: StatusModule dependencies initialized
1/16/2021, 1:56:14 AM - info - RoutesResolver: EntitiesController {/entities}:
1/16/2021, 1:56:14 AM - info - RouterExplorer: Mapped {/entities, GET} route
1/16/2021, 1:56:14 AM - info - RoutesResolver: StatusController {/status}:
1/16/2021, 1:56:14 AM - info - RouterExplorer: Mapped {/status, GET} route
1/16/2021, 1:56:14 AM - info - HomeAssistantService: Successfully connected to MQTT broker at mqtt://192.168.50.100:1883
1/16/2021, 1:56:14 AM - info - ConfigService: Loading configuration from /usr/lib/node_modules/room-assistant/dist/config/definitions/default.js, config/default.json, config/local.json
1/16/2021, 1:56:14 AM - info - ClusterService: Starting mDNS advertisements and discovery
1/16/2021, 1:56:14 AM - info - NestApplication: Nest application successfully started
/usr/lib/node_modules/room-assistant/node_modules/@abandonware/noble/lib/hci-socket/hci.js:100
    this._deviceId = this._socket.bindRaw(deviceId);

Error: ENODEV, No such device
    at Hci.init (/usr/lib/node_modules/room-assistant/node_modules/@abandonware/noble/lib/hci-socket/hci.js:100:35)
    at NobleBindings.init (/usr/lib/node_modules/room-assistant/node_modules/@abandonware/noble/lib/hci-socket/bindings.js:78:13)
    at /usr/lib/node_modules/room-assistant/node_modules/@abandonware/noble/lib/noble.js:60:24
    at processTicksAndRejections (internal/process/task_queues.js:79:11) {
  errno: 19,
  code: 'ENODEV',
  syscall: 'bind'
}
[cmd] /room-assistant/entrypoint.sh exited 1
[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
[s6-finish] sending all processes the TERM signal.
```

I have tried   
```
hciDeviceId:
    - 0
```
for both BTC and BLE at the same time, and I have tried the hci for BTC: 0 and BLE: 1, and I then tried BTC: 1 and BLE: 0, which is what the config currently shows.

**Comment by maxlyth** [comment]
The hciDeviceId parameter is only useful if you have multiple BlueTooth controllers. Unless you have a 3rd party USB BT device plugged into your Pi Zero then there is only one controller and the ID is 0. Trying to use another ID is likely to yield the error you are seeing.

The Room-Assistant documentation does state that it is possible (with performance impact) to use Classic and BLE on the same controller but given that the Zero-W also shares the BT data bus with the WiFi controller then trying to use BLE with Classic is likely to make the obvious bus contention problems significantly worse.

I am facing similar issue and have tried to use a secondary USB BT controller for BLE but have faced bugs (introduced summer 2020) in the core Raspian distro that have broken this setup and I have posted a fix in another thread. Even with this fix I am facing reliability issues and I think the shared UART with WiFi is a likely cause.

I have some Banana Pi M2's on order (a modernised Pi Zero) to see if that improves the situation but for now I have cut back my expectations of the Pi Zero to the bare basics.

**Reply by mKeRix** [reply]
I would love to hear your experience with the Banana Pi M2 Zero! If it ends up being positive I would also like to order some and potentially mention them as a recommendation in the docs.

Other than that, for the performance problems on the Pi Zero there is an open thread: #351. It shouldn't be as bad as it is right now, I've been trying to debug that alongside another topic for a couple weeks now. Once we solve that the limitations should mostly be on BT Classic, as you will need a slightly higher interval to not pound the shared chip too much.

**Comment by mKeRix** [comment]
Can you post the output that you receive when you run `hciconfig` on your Pi?

**Comment by toreupfeet** [comment]
Oh man. I apologize for bringing noob issues to the discussion. Long story short, I was confused as to how the pi zero communicated with home assistant. The log I provided was from my HA add-on (not the Pi) and my PC has a Bluetooth dongle that, I just realized, works only with Windows. So, the HA add-on had no bluetooth receiver for my debian vm. 

I have it running properly on my pi zero now, and MQTT discovered my devices. I'll keep my mouth shut for a bit and try to figure things out before causing any more trouble around here. I have a Pi4 that I'll try this out on as well. 

I see that Room Assistant sees that I'm Home in my Lovelace card. This is great.

**Reply by mKeRix** [reply]
No worries! I and many others in this community are happy to help :)
